oci/archive/oci_dest.go

```diff
@@ -0,0 +1,188 @@
+	return d.unpackedDest.MustMatchRuntimeOS()
+}
+
+// PutBlob writes contents of stream and returns data representing the result (with all data filled in).
+// inputInfo.Digest can be optionally provided if known; it is not mandatory for the implementation to verify it.
+// inputInfo.Size is the expected length of stream, if known.
+func (d *ociArchiveImageDestination) PutBlob(stream io.Reader, inputInfo types.BlobInfo) (types.BlobInfo, error) {
+	valBlobInfo, err := d.unpackedDest.PutBlob(stream, inputInfo)
+	if err != nil {
+		if err := deleteTempDir(); err != nil {
+			return types.BlobInfo{}, errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return types.BlobInfo{}, err
+	}
+	return valBlobInfo, nil
+}
+
+// HasBlob returns true iff the image destination already contains a blob with the matching digest which can be reapplied using ReapplyBlob
+func (d *ociArchiveImageDestination) HasBlob(info types.BlobInfo) (bool, int64, error) {
+	valBool, valInt, err := d.unpackedDest.HasBlob(info)
+	if err != nil {
+		if err := deleteTempDir(); err != nil {
+			return false, -1, errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return false, -1, err
+	}
+	return valBool, valInt, nil
+}
+
+func (d *ociArchiveImageDestination) ReapplyBlob(info types.BlobInfo) (types.BlobInfo, error) {
+	valBlobInfo, err := d.unpackedDest.ReapplyBlob(info)
+	if err != nil {
+		if err := deleteTempDir(); err != nil {
+			return types.BlobInfo{}, errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return types.BlobInfo{}, err
+	}
+	return valBlobInfo, nil
+}
+
+// PutManifest writes manifest to the destination
+func (d *ociArchiveImageDestination) PutManifest(m []byte) error {
+	err := d.unpackedDest.PutManifest(m)
+	if err != nil {
+		if err := deleteTempDir(); err != nil {
+			return errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return err
+	}
+	return nil
+}
+
+func (d *ociArchiveImageDestination) PutSignatures(signatures [][]byte) error {
+	err := d.unpackedDest.PutSignatures(signatures)
+	if err != nil {
+		if err := deleteTempDir(); err != nil {
+			return errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return err
+	}
+	return nil
+}
+
+// Commit marks the process of storing the image as successful and asks for the image to be persisted
+// after the directory is made, it is tarred up into a file and the directory is deleted
+func (d *ociArchiveImageDestination) Commit() error {
+	if err := d.unpackedDest.Commit(); err != nil {
+		if err := deleteTempDir(); err != nil {
+			return errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return errors.Wrapf(err, "error storing image %q", d.ref.image)
+	}
+
+	// path of directory to tar up
+	src := d.tempDir
+	// path to save tarred up file
+	dst := d.ref.resolvedFile
+	if err := tar(src, dst); err != nil {
```

Comment:
(This works just fine; I can’t shake a feeling that _maybe_ `tar` could be renamed to something that makes the semantics of its arguments explicit just by reading the call, and then the `src`/`dst` assignments and the comments could be unnecessary. But I can’t think of a good alternative name for `tar`; `tarDirectoryTo` feels awkward and insufficient. Perhaps someone smarter than me can find a better name; if not, this is quite readable as is.)